src/particle.f90

```diff
@@ -0,0 +1,2351 @@
+
+    allocate(particles(1:nump_read))
+
+    do j=1,nump_read
+      !
+      call particles(j)%init()
+      !
+      particles(j)%x(1:3)=xp(1:3,j)
+      !
+      particles(j)%v(1) = 0._mytype
+      particles(j)%v(2) = 0._mytype
+      particles(j)%v(3) = 0._mytype
+      !
+      particles(j)%new  = .false.
+      !
+    enddo
+
+    deallocate(xp)
+
+    call partical_domain_check(particles)
+
+    call partical_swap(particles)
+
+    call particle_array_clear(particles)
+
+  end subroutine particle_read_bin
+
+  !+-------------------------------------------------------------------+
+  !| This subroutine is to get the fluids velocity on particles.       |
+  !+-------------------------------------------------------------------+
+  !| CHANGE RECORD                                                     |
+  !| -------------                                                     |
+  !| 15-11-2021  | Created by J. Fang                                  |
+  !+-------------------------------------------------------------------+
+  subroutine field_var(ux1,uy1,uz1,B)
+    !
+    use MPI
+    use param,     only : dx,dy,dz,istret,nclx,ncly,nclz,xlx,yly,zlz
+    use variables, only : yp,ny,nz
+    use decomp_2d, only : xsize,xstart,xend,update_halo
+    use actuator_line_model_utils
+    !
+    real(mytype),dimension(xsize(1),xsize(2),xsize(3)),intent(in) :: ux1,uy1,uz1
+    real(mytype),dimension(xsize(1),xsize(2),xsize(3),1:3),intent(in),optional :: B
+    !
+    ! local data
+    integer :: jpart,npart,i,j,k,psize
+    type(partype),pointer :: pa
+    real(mytype) :: x1,y1,z1,x2,y2,z2
+    real(mytype) :: test(xsize(1),xsize(2),xsize(3))
+    real(mytype),allocatable,dimension(:,:,:) :: ux1_halo,uy1_halo,    &
+                                                 uz1_halo,ux1_hal2,    &
+                                                 uy1_hal2,uz1_hal2
+    real(mytype),allocatable :: b_halo(:,:,:,:)
+    !
+    real(mytype),allocatable,save :: xx(:),yy(:),zz(:)
+    logical,save :: firstcal=.true.
+    !
+    if(firstcal) then
+      !
+      allocate(xx(0:xsize(1)+1),yy(0:xsize(2)+1),zz(0:xsize(3)+1))
+      !
+      do i=0,xsize(1)+1
+        xx(i)=real(i-1,mytype)*dx
+      enddo
+      do j=0,xsize(2)+1
+        !
+        if(j+xstart(2)-1>ny) then
+          yy(j)=2._mytype*yp(ny)-yp(ny-1)
+        elseif(j+xstart(2)-1<1) then
+          yy(j)=2._mytype*yp(1)-yp(2)
+        else
+          yy(j)=yp(j+xstart(2)-1)
+        endif
+        !
+      enddo
+      !
+      if(xsize(3)==1) then
+        zz(:)=0._mytype
+      else
+        do k=0,xsize(3)+1
+          zz(k)=real((k+xstart(3)-2),mytype)*dz
+        enddo
+      endif
+      !
+      firstcal=.false.
+      !
+    endif
+    !
+    allocate( ux1_halo(1:xsize(1)+1,0:xsize(2)+1,0:xsize(3)+1),        &
+              uy1_halo(1:xsize(1)+1,0:xsize(2)+1,0:xsize(3)+1),        &
+              uz1_halo(1:xsize(1)+1,0:xsize(2)+1,0:xsize(3)+1) )
+    ! 
+    call pswap_yz(ux1,ux1_halo)
```

Comment:
Is it possible to use halo subroutines available in 2decomp ?

Reply:
Any update on this topic ?

Reply:
yes. pswap_yz uses update_halo from decomp2d.